sygnal/webpushpushkin.py

```diff
@@ -0,0 +1,225 @@
+            with PENDING_REQUESTS_GAUGE.track_inprogress():
+                await self.connection_semaphore.acquire()
+
+        try:
+            with SEND_TIME_HISTOGRAM.time():
+                with ACTIVE_REQUESTS_GAUGE.track_inprogress():
+                    logger.info("sending payload %s", data)
+                    response_wrapper = webpush(
+                        subscription_info=subscription_info,
+                        data=data,
+                        vapid_private_key=self.vapid_private_key,
+                        vapid_claims=self.vapid_claims,
+                        requests_session=self.http_agent_wrapper,
+                    )
+                    response = await response_wrapper.deferred
+                    await readBody(response)
+        finally:
+            self.connection_semaphore.release()
+
+        # assume 4xx is permanent and 5xx is temporary
+        if 400 <= response.code < 500:
+            return [device.pushkey]
+        return []
+
+    @staticmethod
+    def _build_payload(n, device):
+        """
+        Build the payload data to be sent.
+        Args:
+            n: Notification to build the payload for.
+            device (Device): Device information to which the constructed payload
+            will be sent.
+
+        Returns:
+            JSON-compatible dict
+        """
+        payload = {}
+
+        default_payload = device.data.get("default_payload")
+        if isinstance(default_payload, dict):
+            payload.update(default_payload)
+
+        for attr in [
+            "room_id",
+            "room_name",
+            "room_alias",
+            "membership",
+            "event_id",
+            "sender",
+            "sender_display_name",
+            "user_is_target",
+            "type",
+            "content",
+        ]:
+            if getattr(n, attr, None):
+                payload[attr] = getattr(n, attr)
+
+        if getattr(n, "counts", None):
+            counts = n.counts
+            for attr in ["unread", "missed_calls"]:
+                if getattr(counts, attr, None) is not None:
+                    payload[attr] = getattr(counts, attr)
+
+        return payload
+
+
+class HttpAgentWrapper:
+    def __init__(self, http_agent):
+        self.http_agent = http_agent
+
+    def post(self, endpoint, data, headers, timeout):
+        logger.info("HttpAgentWrapper: POST %s", endpoint)
+        body_producer = FileBodyProducer(BytesIO(data))
+        headers = {
+            b"User-Agent": ["sygnal"],
+            b"Content-Encoding": [headers["content-encoding"]],
+            b"Authorization": [headers["authorization"]],
+            b"TTL": [headers["ttl"]],
+        }
+        deferred = self.http_agent.request(
+            b"POST",
+            endpoint.encode(),
+            headers=Headers(headers),
+            bodyProducer=body_producer,
+        )
+        return HttpResponseWrapper(deferred)
+
+
+class HttpResponseWrapper:
```

Comment:
This could also use a docstring, it seems to provide a status code < 202 to avoid error checking in in pywebpush.

I'm not sure why text is provided?

Reply:
text is references by [pywebpush when it logs the response](https://github.com/web-push-libs/pywebpush/blob/main/pywebpush/__init__.py#L373).

Reply:
Thanks! I missed that. 😄